Review

jayrbolton opened the pull request:
Using the tricks from [this blog post](http://astrofrog.github.io/blog/2013/04/10/how-to-conduct-a-full-code-review-on-github/) to compare an empty branch with a review branch

Notable files to review:  `lib/KBaseReportPy/utils/*.py`

jayrbolton committed:
Init and add code for uploading report data with DataFileUtil

jayrbolton committed:
Get most everything working, both simple and extended with validations

jayrbolton committed:
Add a lot of tests around uploading files for html_links or file_links in the extended report

jayrbolton committed:
Add documentation

jayrbolton committed:
Remove test skips

jayrbolton committed:
Correct typo in KBaseReportPy_server_test

jayrbolton committed:
Add more test cases

jayrbolton committed:
Add a test of the resulting report object in ws for the simple method

jayrbolton committed:
Add in the ui spec json file from the original KBaseReport

jayrbolton committed:
Add more test cases; delete junk files; clean up some type validation

jayrbolton committed:
Update README.md

jayrbolton committed:
Merge branch 'master' into review

jayrbolton committed:
Update RELEASE_NOTES.md for 0.0.1

jayrbolton committed:
Remove and ignore .pyc files

jayrbolton committed:
Add utf-8 encoding

MrCreosote commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+# RUN apt-get update
+
+# Update security deps
```
All this security stuff should be taken care of in kbase:sdbase2, can be removed

MrCreosote commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+    cffi pyopenssl ndg-httpsclient pyasn1 requests 'requests[security]'
+
+# Install pip deps
+COPY requirements.txt /kb/module/requirements.txt
```
Why not just pip install the one thing in the requirements doc? Seems overly complicated for something that's probably not going to change.

MrCreosote commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+# Install pip deps
+COPY requirements.txt /kb/module/requirements.txt
+WORKDIR /kb/module
```
WORKDIR is after the chmod in the template (https://github.com/kbase/kb_sdk/blob/master/src/java/us/kbase/templates/module_dockerfile.vm.properties), not sure why it's up here.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,141 @@
+*/
+module KBaseReportPy {
+    /* @id ws */
+    typedef string ws_id;
```
Document

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,141 @@
+    } WorkspaceObject;
+
+    /*
+        Represents a file or html archive that the report should like to
```
truncated?

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,141 @@
+/*
```
All keys in structures in this file should be described. See https://ci.kbase.us/services/ws/docs/Workspace.html for examples.

MrCreosote replied:
Importantly, which fields are required and optional needs to be documented.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,141 @@
+
+    /*
+        A simple Report of a method run in KBase.
+        It only provides for now a way to display a fixed width text output summary message, a
```
This sentence could use a rewrite.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,141 @@
+/*
+    Module for a simple WS data object report type.
+*/
+module KBaseReportPy {
```
This is going to be a real pain, but I assume @scanon will want to rename this to KBaseReport and do DB surgery to make this version replace the old version. That means everywhere in the code base the name will need to be changed, we'll need to rename the old repo, and then push this code into a new repo with the same name as the old repo.

MrCreosote replied:
Copying relevant comment from previous review:

This is going to be a real pain, but I assume @scanon will want to rename this to KBaseReport and do DB surgery to make this version replace the old version. That means everywhere in the code base the name will need to be changed, we'll need to rename the old repo, and then push this code into a new repo with the same name as the old repo.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,141 @@
+    typedef string ws_id;
+
+    /*
+        Reference to a handle
```
It's really a handle id.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,141 @@
+        @metadata ws length(text_message) as Size(characters)
+        @metadata ws length(objects_created) as Objects Created
+    */
+    typedef structure {
```
Does the create() method support all these fields? I thought it didn't.

It looks like the create_extended() method input is a flattened version of create() with a few extra fields.

jayrbolton replied:
The differences between create() and create_extended() are strange, yeah. They both allow all the same data ultimately, but create() is lower level. In particular, create_extended() abstracts the file_links and html_links. 

If you want to look at the perl for create(), it starts at this line and is pretty short; https://github.com/kbaseapps/KBaseReport/blob/master/lib/KBaseReport/KBaseReportImpl.pm#L305

Both functions ultimately create a workspace object that can contain LinkedFile typed objects. The difference is that create() doesn't do anything with the linked files that you pass in -- it just passes them along and uploads them raw.

In my opinion create() should be deleted or completely replaced as it's v confusing

jayrbolton committed:
Add a test case for passing in the workspace ID instead of name

jayrbolton committed:
Ignore the kb-sdk-generated backup files

jayrbolton committed:
Use single underscore function names for "private" functions

MrCreosote commented on .dockerignore:
```diff
@@ -0,0 +1,2 @@
+test_local
```
General comment: I haven't verified that the behavior here matches the behavior of the old module. I'm trusting Jay and some time in CI to do that. @scanon

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,141 @@
+            workspace_id: id
+        }
+    */
+
```
Lose the blank line to make it clear that the docs belong to this method

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,141 @@
+        string message - simple text message to store in report object
+        list <WorkspaceObject> objects_created;
+        list <string> warnings - a list of warning messages in simple text
+        The following argument allows the user to specify the location of html files/directories that the report widget will render <or> link to:
```
100 char per line max

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,141 @@
+        The following argument allows the user to specify the location of files that the report widget should link for download:
+        list <fileRef> file_links - a list of paths or shock node IDs pointing to a single flat file
+        The following parameters indicate where the report object should be saved in the workspace:
+        string report_object_name - name to use for the report object (job ID is used if left unspecified)
```
I don't this this can be true, how would the report code know the job id?

MrCreosote commented on README.md:
```diff
@@ -0,0 +1,80 @@
+
+* `message`: (optional string) basic result message to show in the report
+* `report_object_name`: (optional string) a name to give the workspace object that stores the report (will use job ID if unspecified)
+* `workspace_name`: (required string) string name of your workspace
```
Not required if an id is supplied (and id is strongly recommended since it's non-mutable)

MrCreosote commented on README.md:
```diff
@@ -0,0 +1,80 @@
+* `direct_html`: (optional string) raw HTML to show in the report
+* `objects_created`: (optional list of WorkspaceObject) data objects that were created as a result of running your app, such as assemblies or genomes
+* `warnings`: (optional list of strings) any warnings messages generated from running the app
+* `file_links`: (optional list of dicts) files to attach to the report (see the valid key/vals below)
```
Explain the difference between html and file links

MrCreosote commented on README.md:
```diff
@@ -0,0 +1,80 @@
+* `name`: (required string) name of the file
+* `description`: (optional string) Readable description of the file
+
+For `html_links` the `path` or `shock_id` can be a zipped directory. If you pass a zipped directory with an `index.html` file into `html_links`, then the report can render that `index.html` directly. If you have images or other files within your HTML directory, you can link to those from within your `index.html`.
```
This is confusing to me - does this mean it can be a directory that has been zipped into a zipfile, or that it can be a directory that *will be* zipped into a zipfile by KBR? I thought KBR would do the latter for you.

The 2nd two sentences are also confusing as to what's going on IMO.

MrCreosote commented on lib/src/us/kbase/kbasereport/CreateExtendedReportParams.java:
```diff
@@ -0,0 +1,279 @@
+
```
These all looks like parts of the java client for the old kbasereport?

MrCreosote replied:
copying relevant comment from previous review:

These all looks like parts of the java client for the old kbasereport?

MrCreosote replied:
In particular it looks like there's 2 copies of the java client

MrCreosote replied:
Same comment here - this looks like cruft from the old report.

MrCreosote commented on lib/KBaseReportPy/utils/file_utils.py:
```diff
@@ -0,0 +1,59 @@
+                'make_handle': 1,
+                'pack': 'zip' if zip else None
+            })
+        elif 'shock_id' in file:
```
What if both are missing?

MrCreosote commented on lib/KBaseReportPy/utils/file_utils.py:
```diff
@@ -0,0 +1,59 @@
+    :returns: list of file dictionaries that that can be uploaded to the workspace for the report
+    """
+    out_files = []
+    for file in files:
```
`file` masks a keyword, which doesn't matter here, but generally isn't great practice. My linter complains about it anyway.

MrCreosote commented on lib/KBaseReportPy/utils/file_utils.py:
```diff
@@ -0,0 +1,59 @@
+            shock = dfu.file_to_shock({
+                'file_path': file['path'],
+                'make_handle': 1,
+                'pack': 'zip' if zip else None
```
`zip` also masks a keyword

MrCreosote commented on lib/KBaseReportPy/utils/file_utils.py:
```diff
@@ -0,0 +1,59 @@
+    for file in files:
+        if 'path' in file:
+            # Having a 'path' key means we have to upload to shock
+            shock = dfu.file_to_shock({
```
There's a `file_to_shock_mass` method that could speed things up for multiple uploads, but probably not necessary until someone complains.

MrCreosote commented on lib/KBaseReportPy/utils/file_utils.py:
```diff
@@ -0,0 +1,59 @@
+        elif 'shock_id' in file:
+            # Having a 'shock_id' means it is already uploaded
+            shock = dfu.own_shock_node({'shock_id': file['shock_id'], 'make_handle': 1})
+        out_files.append(_get_file_data(shock, file))
```
This method says it returns Files but _get_file_data says it returns LinkedFiles

MrCreosote commented on lib/KBaseReportPy/utils/file_utils.py:
```diff
@@ -0,0 +1,59 @@
+            )
+
+
+def _get_file_data(shock, file):
```
I would probably just inline this unless there's a good reason not to. Hardly does anything.

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+    :param dfu: instance of DataFileUtil
+    :returns: report data
+    """
+    report_name = "report_" + str(uuid4())
```
Seems like this should throw an error if unsupported fields are included (*_links for instance)

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+        'file_links': [],
+        'html_links': [],
+        'direct_html': '',
+        'direct_html_link_index': 0
```
Will this cause the UI to complain that there is no index 0?

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+            'name': report_name,
+            'meta': {},
+            'hidden': 1,
+            'provenance': ctx['provenance']
```
Don't include the provenance - DFU will handle it for you.

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+    """
+    Create a simple report
+    :param ctx: context dict passed into KBaseReportImpl#create
+    :param params: see utils/validation_utils
```
This line doesn't make sense to me? The params are defined in the spec file.

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+            'provenance': ctx['provenance']
+        }]
+    }
+    obj = dfu.save_objects(save_object_params)[0]
```
I usually wrap calls in try/catch and log the "server side" exception, because otherwise it gets lost. See kb_quast for an example.

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+    :param dfu: instance of DataFileUtil
+    :returns: uploaded report data
+    """
+    validate_paths('file_links', params.get('file_links', []))
```
It looks like validate_paths expects the input to be regular files, but shouldn't directories be allowed as well?

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+    files = fetch_or_upload_files(dfu, params.get('file_links', []))  # see ./file_utils.py
+    html_files = fetch_or_upload_files(dfu, params.get('html_links', []), zip=True)
+    report_data = {
+        'text_message': params.get('message', ''),
```
Better to leave as None as default unless the UI expects and empty string.

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+        'warnings': [],
+        'file_links': [],
+        'html_links': [],
+        'direct_html': '',
```
Better to leave as None unless the UI expects an empty string.

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+        'text_message': params.get('message', ''),
+        'file_links': files,
+        'html_links': html_files,
+        'direct_html': params.get('direct_html', ''),
```
Better to leave as None unless the UI expects an empty string.

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+        'file_links': files,
+        'html_links': html_files,
+        'direct_html': params.get('direct_html', ''),
+        'direct_html_link_index': params.get('direct_html_link_index', 0),
```
If this is zero rather than null and the array is empty will that confuse the ui?

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+            'name': report_name,
+            'meta': {},
+            'hidden': 1,
+            'provenance': ctx['provenance']
```
No provenenance

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+            'provenance': ctx['provenance']
+        }]
+    }
+    obj = dfu.save_objects(save_object_params)[0]
```
wrap try/catch

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+    }
+    obj = dfu.save_objects(save_object_params)[0]
+    ref = _get_object_ref(obj)
+    return {'ref': ref, 'name': report_name, 'shock_id': 'xyz'}
```
Why is an invalid shock id in the return?

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,98 @@
+    if 'workspace_name' in params:
+        return dfu.ws_name_to_id(params['workspace_name'])
+    else:
+        return params['workspace_id']
```
Possible key error here - better to throw an error with a more helpful message

MrCreosote commented on lib/KBaseReportPy/utils/validation_utils.py:
```diff
@@ -0,0 +1,83 @@
+# -*- coding: utf-8 -*-
+from voluptuous import Schema, Required, All, Length, Url
```
This is neat. I've been half-heartedly looking for something similar for java

MrCreosote commented on lib/KBaseReportPy/utils/validation_utils.py:
```diff
@@ -0,0 +1,83 @@
+    'shock_id': basestring,
+    'path': basestring,
+    'description': basestring
+}
```
I'll review tests later - I think this is enough to work on for now.

MrCreosote reviewed (changes requested):
The biggest issue I see is that create() allows the input of LinkedFiles but then, AFAICT, ignores them. If the Perl version does the same thing the API should be cleaned up so it matches reality.

jayrbolton commented:
@MrCreosote thanks! Will work through everything today.

jayrbolton commented:
Regarding the comments on the KIDL spec: that file is just copy-and-pasted over from the old repo -- I avoided touching it. I think it would be worth it for me to clean up a lot of the writing in the comments there (eg. many of the params are misnamed)

jayrbolton committed:
Simplify Dockerfile and pip installation

jayrbolton committed:
Update the comments, annotations, and formatting for the KBaseReportPy.spec KIDL file

jayrbolton committed:
Add required arguments notes to the KIDL spec

jayrbolton committed:
Update README.md

jayrbolton committed:
Add a validation case for every html_link or file_link to check for either shock_id or path

jayrbolton committed:
Clean up some validations and default data setting based on MrCreosote's suggestions

jayrbolton committed:
Add try/catch wrapping around calls to dfu.save_objects

jayrbolton committed:
Simplify some of the zipping logic in file_utils

jayrbolton committed:
All html_links must be zipped -- fix fetch_or_upload_files to reflect that

jayrbolton committed:
Clean up some parameters in the report_utils module

jayrbolton committed:
Improve comment

jayrbolton committed:
Add a validation check on the existence of the main HTML for html_links, along with tests

jayrbolton committed:
Add note about main html file

jayrbolton committed:
Add improvements to the run-time type validations and type error messages, making them human-readable

jayrbolton committed:
Fix validation schema on objects_created

jayrbolton committed:
Correct package name

jayrbolton committed:
Correct package name; update documentation

jayrbolton committed:
Add support for handling single HTML files along with a test

jayrbolton committed:
Merge branch 'master' of https://github.com/jayrbolton/KBaseReportPy

jayrbolton commented:
@MrCreosote this is ready for a re-review. See the change notes from my comment above. Since then, I also did some additional improvements based on testing in a narrative:
* Improved validation and error message formatting in `validation_utils.py` (see `validation_utils._format_errors`)
* Validating existence of the main HTML file in the HTML report -- ie you set something like `index.html` as the name in `html_links`, and we make sure that file exists. 
* Handles passing a single file for an html report, rather than a directory
* If any hash in `file_links` has a `path` key that points to a directory, we zip that directory

MrCreosote commented on .travis.yml:
```diff
@@ -0,0 +1,31 @@
+  - git clone https://github.com/kbase/kb_sdk
+  - cd kb_sdk
+  - make bin
+  - make sdkbase
```
With the sdkbase2 change is this needed?

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+/*
+    Module for a simple WS data object report type.
```
Define acronyms

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+     * that can be associated with the object.
+     * Required arguments:
+     *     ws_id ref - workspace ID
+     * @optional description
```
The optional arguments' semantics should be defined as well

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+     * Required arguments:
+     *     handle_ref handle - Handle ID
+     *     string name - Plain-text name of the file (shown to the user)
+     *     string URL - shock URL and ID (`shock['handle']['url'] + '/node/' + shock['handle']['id']`)
```
is this definitely a shock url? I'm not sure.

Edit: it must be.

MrCreosote replied:
Per a different comment, we'll see if we can get rid of LinkedFile altogether

MrCreosote replied:
We can't because it's a part of the Report type, which is registered in the workspace.

MrCreosote replied:
`shock['handle']['url'] + '/node/' + shock['handle']['id']` - this is really implementation specific.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+     * create_extended() is the preferred method if you have html_links and
+     * file_links, but this is still provided for backwards compatibility.
+     */
+    funcdef create(CreateParams params)
```
Mark as deprecated
https://ci.kbase.us/services/ws/docs/typedobjects.html#deprecated-annotation

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+     * Provides a fixed-width, text-based summary message, a list of warnings,
+     * and a list of objects created.
+     * Required arguments:
+     *     string text_message - Readable plain-text report message
```
Looks like objects_created is required as well

jayrbolton replied:
In the python, we default objects_created to `[]` if they don't set it. In that case, would you still want it listed as required in these docs? Maybe it's required in the workspace, but it's not required as a parameter to the app.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+     * Parameters for the create() method
+     * Pass in *either* workspace_name or workspace_id -- only one is needed
+     * Required arguments:
+     *    int workspace_id - needed if workspace_name is blank. Preferred as its immutable.
```
it's

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+     * The reference to the saved KBaseReport. This is the return object for
+     * both create() and create_extended()
+     * Required arguments:
+     *    ws_id ref
```
Define what these are (see https://ci.kbase.us/services/ws/docs/Workspace.html#typedefWorkspace.ObjectSaveData for an example of acceptable documentation)

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+     * @optional warnings file_links html_links direct_html direct_html_link_index
+     * @metadata ws length(warnings) as Warnings
+     * @metadata ws length(text_message) as Message Length
+     * @metadata ws length(objects_created) as Objects Created
```
As we discussed, see if we can remove everything other than message, warnings, and objects_created, since the behavior of create() is so weird. Check that no apps are using those fields, and if not, ping the implementation team channel about it.

MrCreosote replied:
Can't - part of the Report type registered in the workspace.

MrCreosote replied:
maybe 2 different specs, one for the Report type (registered with the Workspace) and one for the API (registered with the Catalog).

`#include` could be useful to reuse Report types in the API: http://htmlpreview.github.io/?https://github.com/kbaseapps/ReadsUtils/blob/master/ReadsUtils.html

jayrbolton replied:
How does `#include` work? is it importing a file locally or some other way?
Is there a way to move `Report` and `LinkedFile` into a separate spec file, and then `#include` it into `KBaseReportPy.spec`?

I made the change to have `create()` use a new, reduced `SimpleReport` datatype for the api, but all types are still in one file.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+    /*
+     * The reference to the saved KBaseReport. This is the return object for
+     * both create() and create_extended()
+     * Required arguments:
```
Since this is a returned data structure, not an input, specifying required arguments doesn't make sense

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+        returns (ReportInfo info) authentication required;
+
+    /*
+     * A file to be linked or uploaded for the report. In extended_report(),
```
No need to expose implementation details

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+     * @metadata ws length(objects_created) as Objects Created
+     * @metadata ws length(html_links) as HTML Links
+     * @metadata ws length(file_links) as File Links
+     * @optional message objects_created warnings html_links direct_html direct_html_link_index file_links report_object_name html_window_height summary_window_height
```
Split into 2 lines

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+     * int direct_html_link_index - Index in html_links to set the direct/default view in the
+     *     report (ignored if direct_html is present).
+     * string direct_html - simple html text that will be rendered within the report widget
+     *     (do not use both this and html_links -- use one or the other)
```
Will an error be thrown if both are passed in?

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+     *     (will be auto-generated if unspecified)
+     * html_window_height - fixed height in pixels of the html window for the report
+     * summary_window_height - fixed height in pixels of the summary window for the report
+     * string workspace_name - name of workspace where object should be saved
```
Are these two arguments exclusive? Is at least one required?

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+/*
```
Note that the metadata and optional parameters only actually do anything for workspace types. The only type registered in the workspace is Report (note LinkedFile is a dependency)

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+
+    /*
+     * A simple Report of a method run in KBase.
+     * Provides a fixed-width, text-based summary message, a list of warnings,
```
Description is out of date

jayrbolton replied:
Not sure which part you mean. The fixed-width bit?

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+     * Required arguments:
+     *    int workspace_id - needed if workspace_name is blank. Preferred as its immutable.
+     *    string workspace_name - needed if workspace_id is blank. Note that this may change.
+     *    Report report
```
In order to avoid the weird behavior of create(), maybe make a MinimalReport (or something) structure that only has the message, warnings, and ws_objs keys and use that here instead of report..

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,171 @@
+    /*
+     * A file to be linked or uploaded for the report. In extended_report(),
+     * this will get converted into a lower-level LinkedFile before uploading
+     * Pass in *either* a shock_id or a path. If a path is given, then the file
```
Will it zip directories?

MrCreosote commented on LICENSE:
```diff
@@ -0,0 +1,7 @@
+Copyright (c) 2015 The KBase Project and its Contributors
```
2015-2018

MrCreosote commented on README.md:
```diff
@@ -0,0 +1,83 @@
+Use the method **`report_client.create_extended_report(params)`** to create a report object along with the following parameters, passed as a dictionary:
+
+* `message`: (optional string) basic result message to show in the report
+* `report_object_name`: (optional string) a name to give the workspace object that stores the report (will use job ID if unspecified)
```
I don't think this is true, the code doesn't know the job id

MrCreosote commented on lib/KBaseReportPy/utils/file_utils.py:
```diff
@@ -0,0 +1,76 @@
+                'make_handle': 1,
+                'pack': 'zip' if isdir else None
+            })
+        elif 'shock_id' in each_file:
```
What if neither path or shock_id are present?

jayrbolton replied:
This will also be caught by `validation_utils` further up in the execution (see `_validate_files`)

MrCreosote replied:
This seems odd to me. The functions in the utils packages often seem to depend on the user calling a validate function on the inputs prior to calling the function proper. Wouldn't it make more sense to the function to call the validate method itself? That makes it easier for the user and less error prone, since the function knows for sure that the input has been validated.

MrCreosote commented on lib/KBaseReportPy/utils/file_utils.py:
```diff
@@ -0,0 +1,76 @@
+                os.makedirs(new_dir)
+                # Move the file to dir/name
+                new_path = os.path.join(new_dir, each_file['name'])
+                os.rename(each_file['path'], new_path)
```
This effectively deletes the original file, right? That doesn't seem like a good idea.

MrCreosote commented on lib/KBaseReportPy/utils/file_utils.py:
```diff
@@ -0,0 +1,76 @@
+                'make_handle': 1,
+                'pack': 'zip'  # Always zip for HTML
+            })
+        elif 'shock_id' in each_file:
```
What if nieither path nor shock_id are in each_file?

jayrbolton replied:
This case is handled in the `validation_utils` (the `_validate_files` function). Will raise an error further up the execution

MrCreosote commented on lib/KBaseReportPy/utils/file_utils.py:
```diff
@@ -0,0 +1,76 @@
+    """ This corresponds to the LinkedFile type in the KIDL spec """
+    return {
+        'handle': shock['handle']['hid'],
+        'description': f.get('description', ''),
```
Does the UI expect an empty string or None? I'd assume None but I don't know.

MrCreosote commented on lib/KBaseReportPy/utils/report_utils.py:
```diff
@@ -0,0 +1,103 @@
+    if 'workspace_name' in params:
+        return dfu.ws_name_to_id(params['workspace_name'])
+    else:
+        return params.get('workspace_id')
```
may return null

jayrbolton replied:
`validation_utils` makes sure that workspace_id is an integer (min 0) if it is present, or workspace_name is a string (minlength 1) if present.1

MrCreosote commented on lib/KBaseReportPy/utils/validation_utils.py:
```diff
@@ -0,0 +1,191 @@
+        pprint.pformat(params),
+        "\n(View the type spec here: ",
+        # TODO update this URL when it moves and is final:
+        "https://github.com/jayrbolton/KBaseReportPy/blob/master/KBaseReportPy.spec)\n",
```
could compile the spec to HTML and link to that instead, like this: http://htmlpreview.github.io/?https://github.com/kbaseapps/ReadsUtils/blob/master/ReadsUtils.html

MrCreosote replied:
although I'm not sure that linking out to a url in an error message is a great idea. Maybe just say check the docs or something like that.

jayrbolton replied:
I personally really like to see help URLs in error messages. But I agree it doesn't make sense if the URL is too ephemeral. I'll remove for now.

jayrbolton committed:
Fix comment to be non-implementation-specific

jayrbolton committed:
Remove unnecessary travis build command

jayrbolton committed:
Add kb-sdk test as the travis command

jayrbolton committed:
Improve spec module description comment

jayrbolton committed:
Refactor the .travis.yml build modeled after some more modern apps

jayrbolton committed:
Use shutil to move the html index file into a directory

jayrbolton committed:
Comment out the link to the KIDL spec in the error message formatter for now

jayrbolton committed:
Have a test without description key

jayrbolton committed:
Copy the HTML file into a directory instead of moving; Set description to None, not empty-string

jayrbolton committed:
Move the simple create() parameter type into a SimpleReport type

jayrbolton committed:
Don't compile the app client files in the Makefile

jayrbolton committed:
Remove perl/python/java client code

jayrbolton committed:
Run make to compile updated KIDL spec

jayrbolton committed:
Merge branch 'master' into simplify-kidl-spec

jayrbolton committed:
Update year

jayrbolton committed:
Correct comment about generating report name

jayrbolton committed:
Update documentation comments based on review feedback

jayrbolton committed:
Pip install on one line

jayrbolton committed:
Merge branch 'review' into simplify-kidl-spec

jayrbolton committed:
Merge pull request #2 from jayrbolton/simplify-kidl-spec

Simplify kidl spec

jayrbolton committed:
Merge branch 'review'

jayrbolton committed:
Add the direct_html string to the py validation for create_extended_report

jayrbolton committed:
Change shutil.copy -> shutil.copy2

jayrbolton committed:
Try not zipping a single html file to see what happens

jayrbolton committed:
Revert back to wrapping single HTML files in a directory before zipping

jayrbolton commented:
@MrCreosote I had a question related to changing around the kidl spec for this app. I made the update to have simple reports take a `SimpleReport` type that is separate from `Report` and does not contain `file_links` or `html_links`. However, the spec is still all one file. I wasn't sure how to split the files.

> How does #include work? is it importing a file locally or some other way?
Is there a way to move Report and LinkedFile into a separate spec file, and then #include it into KBaseReportPy.spec?

> I made the change to have create() use a new, reduced SimpleReport datatype for the api, but all types are still in one file.

MrCreosote commented:
Example:
https://github.com/kbaseapps/ReadsUtils/blob/master/ReadsUtils.spec

For uploading to the workspace (which you don't need to do here) it's a little different, I think - haven't done that in forever so I don't really remember. Moot here anyway.

jayrbolton committed:
Add full perms on the temp dir created for single HTML files to try to get things to work on Travis

jayrbolton committed:
Split the KIDL spec into an API spec for the catalog and a Workspace spec

jayrbolton committed:
Try out 755 permissions for Travis

jayrbolton committed:
Remove API types from KBaseReportPyWorkspace.spec

jayrbolton commented:
I split the KIDL files into KBaseReportPy.spec (API) and KBaseReportPyWorkspace.spec (Workspace). Also got the tests running on Travis finally

jayrbolton commented:
@MrCreosote This could use another review if you get a chance. Could also do it on this PR here, although the diff will be against the perl codebase: https://github.com/kbaseapps/KBaseReport/pull/8

MrCreosote commented on KBaseReportPyWorkspace.spec:
```diff
@@ -0,0 +1,69 @@
+ * Behind-the-scenes types for report data that gets uses on the workspace. 
+ * These objects are generated in this module from the API types found in KBaseReportPy
+ */
+module KBaseReportPyWorkspace {
```
This will need to be called KBaseReport: https://ci.kbase.us/#spec/module/KBaseReport

MrCreosote commented on .travis.yml:
```diff
@@ -0,0 +1,29 @@
+  - docker run kbase/kb-sdk genscript > $HOME/bin/kb-sdk
+  - chmod 755 $HOME/bin/kb-sdk
+  - export PATH=$PATH:$HOME/bin/kb-sdk
+  - kb-sdk sdkbase
```
Is this still required with sdkbase2?

jayrbolton replied:
This pulls the docker image for kbase/kbase:sdkbase2.latest -- I think it is necessary

MrCreosote commented on .travis.yml:
```diff
@@ -0,0 +1,29 @@
+
+env:
+  global:
+    secure: LYbfSjXXh6i0QxQrlMuR1SzWs/HZux5RMua9kbpB2bpf5rq7a5c/Ztgzm3bEktT7tUEy4Xj3EI5wXivyAyNK8M68xZi3CotuS0cByB/i8s5Y8RT+UaGqOpd9ViwDMXjfrUGH+aFfmWUvhb/P+rmt4/XhUuF1iNnQTfARDJHmwbGyRX7fmzlN0OrajDnaSDa3u1kE6wd6Re5u3HEkkRjp7haiRTG0clsXPs+Ybjc1Da0VDOPw6Ksm7b6RXwrEuxKlVPXwXYROOEYyrO71ZYIc7o7xWZDx507ShqeX4TGNHwPzO00Ko1luG5Qzvc8+7iWA4dKXusX1UXMlzd7qiSBCltG9egdDZhE0U4hqu5m5fpfXUgHG0K1G17d2ds8YuweCxFg ...
```
Document what this is.

MrCreosote commented on .travis.yml:
```diff
@@ -0,0 +1,29 @@
+  - kb-sdk test || true
+  - make build-test-script
+  - sed -i "s/test_token=/test_token=$TEST_TOKEN/" test_local/test.cfg
+  - sed -i 's/$(id -u)/0:0/' test_local/run_tests.sh
```
Document what this is for

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,132 @@
+#include <KBaseReportPyWorkspace.spec>
+/*
+ *  Module for workspace data object reports, which show the results of running a job in an SDK app.
```
This sounds a bit off to me. Maybe something like

Module for creating Report objects that can be used to present rich HTML based results, file downloads, and warnings to a user.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,132 @@
+     *         has created. They will get linked in the report view.
+     *     list<string> warnings - A list of plain-text warning messages
+     *     string direct_html - simple html text that will be rendered within the report widget
+     * @optional text_message warnings objects_created
```
Another reminder that these annotations do nothing in this context. They only actually do anything for types in the workspace.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,132 @@
+    /*
+     * Parameters used to create a more complex report with file and html links
+     *
+     * All parameters are optional.
```
This isn't really true, at least on of ws id or name is required.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,132 @@
+     * int direct_html_link_index - Index in html_links to set the direct/default view in the
+     *     report (ignored if direct_html is present).
+     * string direct_html - simple html text that will be rendered within the report widget
+     *     If you pass in both direct_html and html_links, then direct_html will be ignored.
```
This ignoring is a little weird. If you set the html_links and direct_html then direct_html gets ignored but so does direct_html_link_index. Seems like the latter shouldn't happen.

Or maybe just make setting html_links and direct at the same time throw an error.

jayrbolton replied:
Should we worry about breaking any existing apps? I haven't seen any apps that set both direct_html and html_links, but I'm not sure.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,132 @@
+     * int workspace_id - id of workspace where object should be saved.
+     *     Required if workspace_name is not present.
+     *
+     * @optional message objects_created warnings html_links direct_html direct_html_link_index file_links report_object_name html_window_height summary_window_height
```
Again, this does nothing, but you can split this into multiple lines with multiple @optional annotations.

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,132 @@
+
+    /*
+     * Create a report for the results of an app run -- handles file and html zipping/uploading
+     * If you are using html_links or file_links, this will be more user-friendly than create()
```
If you're using those, you can't use create anymore

MrCreosote commented on KBaseReportPyWorkspace.spec:
```diff
@@ -0,0 +1,69 @@
+/*
+ * Behind-the-scenes types for report data that gets uses on the workspace. 
```
gets used in

MrCreosote replied:
These are only sort of behind the scenes, as they describe the schema of the data objects stored in the workspace which are retrieved by workspace clients. For example, a user could use a code cell in the narrative to pull back a report object stored in the workspace and do whatever with it.

MrCreosote commented on KBaseReportPyWorkspace.spec:
```diff
@@ -0,0 +1,69 @@
+     * Represents a Workspace object with some brief description text
+     * that can be associated with the object.
+     * Required arguments:
+     *     ws_id ref - workspace ID
```
Describe more fully, e.g. needs to be in the X/Y/Z format

MrCreosote commented on KBaseReportPyWorkspace.spec:
```diff
@@ -0,0 +1,69 @@
+     * Required arguments:
+     *     handle_ref handle - Handle ID
+     *     string name - Plain-text name of the file (shown to the user)
+     *     string URL - URL of shock node: <shock-url>/node/<shock-handle-id>
```
Missing description of optional args

MrCreosote commented on KBaseReportPyWorkspace.spec:
```diff
@@ -0,0 +1,69 @@
+
+    /*
+     * A Report of a method run in KBase. This is used behind-the-scenes in the
+     * workspace for the simple create() method
```
This is a little confusing. Both create and create extended save an object of this type to the workspace.

MrCreosote commented on KBaseReportPyWorkspace.spec:
```diff
@@ -0,0 +1,69 @@
+     * A Report of a method run in KBase. This is used behind-the-scenes in the
+     * workspace for the simple create() method
+     * Required arguments:
+     *     string text_message - Readable plain-text report message
```
According to the create extended docs, this is not required. Also, the workspace will accept nulls for primitive types so the workspace doesn't require a value (it will require the key though)

MrCreosote replied:
Missing documentation for most arguments

jayrbolton committed:
Include WorkspaceObject and ws_id inside the API spec and dont import from KBaseReportPyWorkspace.spec

jayrbolton committed:
Merge branch 'master' into review

jayrbolton committed:
Rename the Workspace type spec to KBaseReport

jayrbolton committed:
Add comment about TEST_TOKEN secure var

jayrbolton committed:
Add comments in the travis install setup

jayrbolton committed:
Update .spec documentation based on review feedback

jayrbolton commented:
Ok @MrCreosote, I addressed all your comments -- ready for a re-review

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,145 @@
+
+    /*
+     * Parameters for the create() method
+     * Pass in *either* workspace_name or workspace_id -- only one is needed
```
Note ID is preferred as it's immutable

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,145 @@
+     * Required arguments:
+     *     string workspace_name - Name of the workspace where the report
+     *         should be saved. Required if workspace_id is not present
+     *     int workspace_id - ID of workspace where the report should be saved.
```
Note ID is preferred as it's immutable

MrCreosote commented on KBaseReportPy.spec:
```diff
@@ -0,0 +1,145 @@
+     *         If you pass in paths, they will be zipped and uploaded
+     *     int direct_html_link_index - Index in html_links to set the direct/default view in the
+     *         report (ignored if direct_html is present)
+     *     string direct_html - Simple HTML text that will be rendered within the report widget
```
The ignoring stuff here is still confusing. If taken at face value, if you pass in html_links, direct_html_link_index, and direct_html, then direct_html and direct_html_link_index will be ignored, which is not what you want.

jayrbolton replied:
I had a reply previously on this -- throwing an error could possibly break existing apps. I did not find any apps that set both fields, but it's possible it's out there. Would you rather be stricter with throwing an error for these fields or risk breaking an app somewhere?

MrCreosote replied:
didn't say anything about throwing an error - I'm talking about the nonsensical behavior (at least as documented, not sure if that's the actual behavior). We should at least fix that.

That being said, my preference would be to throw an exeption if the user provides incompatible parameters, but I leave that up for you.

jayrbolton replied:
Ignoring direct_html if direct_html_link_index is present is actual behavior -- I tested it manually in the narrative. I added a note that the user should avoid setting both.

MrCreosote replied:
But if html_links is also present is direct_..._index still ignored? That's the goofy part. If html_links takes precedence over direct_html, direct_html should not be able to blow away the index if html_links is present.

jayrbolton replied:
direct_html doesn't override direct_html_link_index -- it's the other way around. 

I think there might be miscommunication on this one. Maybe we can work it out in person. 
- if both direct_..._index and html_links are set, then direct_html is ignored, whether or not it is set.
- if html_links is empty but direct_..._index is set, then we get an out-of-bounds, as expected
- if direct_html is set, direct_.._index is not set, and html_links is non-empty, then direct_html is rendered, while the html_links are listed for download below (in the narrative)

MrCreosote replied:
Here's where I'm confused then:
```
int direct_html_link_index - Index in html_links to set the direct/default view in the
       report (ignored if direct_html is present)
```

That appears to contradict what you say above.

jayrbolton replied:
Ah that's my mistake. I must have been confused when I made that comment. Thought I had updated all those comments, but maybe I didn't push. Will fix.

MrCreosote commented on KBaseReportPyWorkspace.spec:
```diff
@@ -0,0 +1,85 @@
+     * links to or renders as HTML. Used by Report for either html_links or file_links.
+     * Required arguments:
+     *     handle_ref handle - Handle ID
+     *     string name - Plain-text name of the file (shown to the user)
```
The difference between name and label is not at all clear to me.

jayrbolton replied:
My hunch is name is supposed to be filename (ie. "results.zip") while label is supposed to be a short description (eg. "Annotation results"). But I really don't know. I have experimented with all the fields and have found that:
- neither label nor description show up for file_links
- label does not show up for html_links, but description does (underneath name)

Also, I found that "warnings" does not show up for create_extended_report

MrCreosote replied:
What do you mean by show up? Doesn't show up in the workspace object? Doesn't show up in the narrative display of the workspace object?

MrCreosote replied:
You might want to confab with Erik or Bill to see how the report is displayed.

jayrbolton replied:
By "show up", I meant they don't show in the narrative UI

MrCreosote replied:
Then yeah, probably best to talk to B&E and figure out if there's bugs in the display and what the various fields mean.

How do label and name even get populated?

jayrbolton replied:
It looks like name and description can be set in File for the file_links or html_links fields, but label is not anywhere in KBaseReport.spec. I assume it's a legacy thing.

MrCreosote replied:
Per our conversation Jay is going to double check the Perl code re name and label

MrCreosote replied:
Looks like label is used by kb_quast in create extended: https://github.com/kbaseapps/kb_quast/blob/master/lib/kb_quast/kb_quastImpl.py#L209

It shows up as the link name in a report:
https://ci.kbase.us/narrative/ws.16325.obj.1

MrCreosote commented on KBaseReportPyWorkspace.spec:
```diff
@@ -0,0 +1,85 @@
+     *         lists out for the user to download
+     *     list<LinkedFile> html_links - List of zipped directories of HTML
+     *         files and content that the report will render or link for download
+     *     string direct_html - Simple HTML text that will be rendered within the report widget.
```
Is the ignore stuff relevant here? Surely that's dependent on the application that pulls the report from the workspace, and so it could be different per application. 

Maybe just note that the report *should* contain either html_links and direct_..._index or direct_html, but not both.

MrCreosote commented on KBaseReportPyWorkspace.spec:
```diff
@@ -0,0 +1,85 @@
+        list<LinkedFile> file_links;
+        list<LinkedFile> html_links;
+        string direct_html;
+        int direct_html_link_index;
```
missing html_window_height and summary_window_height

jayrbolton replied:
That's not in the original spec for Report: https://github.com/kbaseapps/KBaseReport/blob/master/KBaseReport.spec#L50

Is it somewhere else?

jayrbolton replied:
It must have just been missing from the original app -- I'm seeing it definitely should belong in there

MrCreosote commented on Makefile:
```diff
@@ -0,0 +1,61 @@
+	kb-sdk compile $(SPEC_FILE) \
+		--out $(LIB_DIR) \
+		--pysrvname $(SERVICE_CAPS).$(SERVICE_CAPS)Server \
+		--pyimplname $(SERVICE_CAPS).$(SERVICE_CAPS)Impl;
```
Add compilation for the HTML spec as well. See DataFileUtil for an example.

MrCreosote replied:
We should probably just add that in the template.

jayrbolton replied:
I added the `--html` flag and ran make  -- It is generating an html and css file directly inside `/lib`. Not sure if that's expected behavior

MrCreosote replied:
Looks like DFU is a bad example and needs an update. This one should work: 
https://github.com/kbaseapps/ReadsUtils/blob/master/Makefile#L30

MrCreosote commented on README.md:
```diff
@@ -0,0 +1,83 @@
+
+# KBaseReportPy
+
+This is a [KBase SDK](https://github.com/kbase/kb_sdk) app that can be used within other apps to generate output reports in the narrative.
```
Strictly speaking any application could pull the report and use it, not just the narrative, although that's going to be the major consumer.

MrCreosote commented on README.md:
```diff
@@ -0,0 +1,83 @@
+
+This is a [KBase SDK](https://github.com/kbase/kb_sdk) app that can be used within other apps to generate output reports in the narrative.
+
+[How to create a KBaseReport](https://github.com/kbase/kb_sdk/blob/master/doc/howto/create_a_report.md)
```
Is this currently or soon to be a dead link?

MrCreosote commented on RELEASE_NOTES.md:
```diff
@@ -0,0 +1,10 @@
+# KBaseReportPy release notes
+=========================================
+
+0.0.1
```
Are there no release notes or versions in the original repo? 

This should be at least 0.2.0 IMO.

MrCreosote commented on lib/KBaseReportPy/utils/validation_utils.py:
```diff
@@ -0,0 +1,183 @@
+    """
+    if (index is None) or (not html_links):
+        return
+    html_link = html_links[index]
```
This could throw an out of bounds exception

jayrbolton committed:
Update link to point to newer gh-pages docs

jayrbolton committed:
Add an out-of-bounds check for direct_html_link_index with a test

jayrbolton committed:
Add release notes from the perl version of KBaseReport

jayrbolton committed:
Tweak wording in README.md

jayrbolton committed:
Add option for --html

jayrbolton committed:
Run make

jayrbolton committed:
Remove junk compilation files

jayrbolton committed:
Add the fields for html_window_height and summary_window_heigth to KBaseReportPyWorkspace.spec

jayrbolton committed:
Edit some wording for the direct_html/direct_html_link_index either/or relationship

jayrbolton committed:
Add html generation for KIDL spec; remove old html/css files in lib/

jayrbolton committed:
Add some explicit notes about workspace_id's being preferred in the KIDL spec

jayrbolton committed:
Run make

jayrbolton committed:
Update KIDL specs based on review feedback

- Add more accurative descriptions around direct_html and
direct_html_link_index
- Small copy tweaks

jayrbolton committed:
Set version to 3.0.0

jayrbolton committed:
Move file validation call into file_utils to make it more clear that it gets run

jayrbolton commented:
Addressed the comments, ready for a re-review. Notes:

* Added the release notes from the original KBaseReport repo (not much there). I set the version as 3.0.0 from 2.3.1
* Added KIDL html output
* Attempted to clean up the doc wording around direct_html and direct_.._index
* Added out-of-bounds check for direct_.._index
* I moved the file validation call into file_utils to make it clearer

jayrbolton committed:
Add the label field into the spec

jayrbolton commented:
Added the label field into the File type in the KBaseReportPy spec

jayrbolton committed:
Add the "warnings" parameter into create_extended_report
